graphprot/NeuralNet.py

```diff
@@ -97,25 +97,65 @@ def __init__(self, database, Net,
             self.loss = nn.CrossEntropyLoss(weight = self.class_weights, reduction='mean')       
 
 
-    def train(self, nepoch=1, validate=False):
+    def train(self, nepoch=1, validate=False, plot=False):
+        
+        _train_acc = []
+        _train_loss = []
+        _valid_acc = []
+        _valid_loss = []
 
-        self.model.train()
         for epoch in range(1, nepoch+1):
+            self.model.train()
             t0 = time()
             acc, loss = self._epoch(epoch)
             t = time() - t0
+            _train_loss.append(loss)
+
             if acc is not None:
+                _train_acc.append(acc)
                 print('Epoch [%04d] : train loss %e | accuracy %1.4e | time %1.2e sec.' % (epoch, loss, acc, t))
             else:
                 print('Epoch [%04d] : train loss %e | accuracy None | time %1.2e sec.' % (epoch, loss, t))
                           
-            if validate:
+            if validate is True:
                 _, val_acc, val_loss = self.eval(self.valid_loader)
                 t = time() - t0
+                _valid_loss.append(val_loss)
+
                 if acc is not None :
-                    print('Epoch [%04d] : valid loss %e | accuracy %1.4e | time %1.2e sec.' % (val_loss, val_acc, t))
+                    _valid_acc.append(val_acc)
+            
+                    print('Epoch [%04d] : valid loss %e | accuracy %1.4e | time %1.2e sec.' % (epoch, val_loss, val_acc, t))
                 else :
-                    print('Epoch [%04d] : valid loss %e | accuracy None | time %1.2e sec.' % (epoch, loss, t))
+                    print('Epoch [%04d] : valid loss %e | accuracy None | time %1.2e sec.' % (epoch, val_loss, t))
+
+        if plot is True :
+
+            import matplotlib.pyplot as plt
+            
+            if len(_valid_loss) > 1:
+                plt.plot(range(1,nepoch+1), _valid_loss, c='red', label='valid')
+
+            if len(_train_loss) > 1:
+                plt.plot(range(1,nepoch+1), _train_loss, c='blue', label='train')
+                plt.title("Loss/ epoch")
+                plt.xlabel("Number of epoch")
+                plt.ylabel("Total loss")
+                plt.legend()
+                plt.savefig('loss_epoch.png')
+                plt.close()
+                
+            if len(_valid_acc) > 1:
+                plt.plot(range(1,nepoch+1), _valid_acc, c='red', label='valid')
+
+            if len(_train_acc) > 1:
+                plt.plot(range(1,nepoch+1), _train_acc, c='blue', label='train')
+                plt.title("Accuracy/ epoch")
+                plt.xlabel("Number of epoch")
+                plt.ylabel("Accuracy")
+                plt.legend()
+                plt.savefig('acc_epoch.png')
+                plt.close() 
```

Comment:
So as much as possible we should try to have each function do exactly 1 thing (aka the single responsibility principle). Now if you put the  plot in train(), train() does 2 thing it trains and it plots. As everyone I'm not the best at respecting the single resp ppl but I can try to when I review code :) 

One option would be to have train() return all the losses that can then be used as input of a plot function. Or have member variable for the losses, or export all that in a hdf5 file that you can read afterwards.